### ISSUE TITLE
[intake] Complete ProjectOdyssey issue #3181: implement the full backward

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ml-odyssey#5510]
## Intake

**Idea:** Complete ProjectOdyssey issue #3181: implement the full backward pass for the ResNet-18 CIFAR-10 training script and validate that it trains an epoch

**Context:** Target repo HomericIntelligence/ProjectOdyssey. Issue #3181 already carries state:plan-go with an approved plan: full backward pass through all residual blocks, batch_norm2d_backward in shared/core/normalization.mojo, activation caching in the forward pass, 84 SGD-momentum velocity tensors, parameter updates, and training-convergence validation on CIFAR-10. Describe the work as at most 3 dependency-ordered tasks for pipeline.task-agent; the final PR must include 'Closes #3181'; validation must produce a training log for one epoch (bounded CIFAR-10 subset acceptable if a full epoch exceeds 60 min on CPU).

Interview transcript and researched brief follow as comments.

**Comment by mvillmow** [comment]
# Implementation Plan

Everything is confirmed. #3181's work is fully delivered, validated (a genuine 17.8-hour real-CIFAR-10 one-epoch log exists), the backward tests now gate CI (#5557), and #3181 is closed. The one-epoch training smoke is tracked separately as #5551.

I now have all facts needed to write the intake plan for #5510. The honest framing: the underlying work is **already complete**; the intake's job is a bounded verify-and-close-out that (a) reproduces one-epoch evidence via a channel the agent doesn't author (ADR-014), and (b) lands a reference PR — using `Refs #3181` since the issue is already closed (a `Closes #3181` on a closed issue is a no-op and would misrepresent state).

# Implementation Plan

## Objective

GitHub issue #5510 is an **intake** wrapping ProjectOdyssey issue #3181 ("implement the full backward pass for the ResNet-18 CIFAR-10 training script and validate one epoch"). Codebase verification shows #3181's work is **already fully delivered and merged**: the full manual backward pass, `batch_norm2d_backward`, activation caching, 82 named SGD-momentum velocity tensors, and per-parameter updates all live in `examples/resnet18_cifar10/train.mojo` (1225 lines; commits `173ecdca`, `8dda697f`, `2a4c12c0`), the backward tests now gate CI (`4381bddb`/#5557), and #3181 is **CLOSED as COMPLETED** (2026-07-05) by merged PR #5546. A genuine 17.8-hour real-CIFAR-10 one-epoch log already exists at `examples/resnet18_cifar10/validation/epoch1.log` (loss 1.79→1.38, 60.02% test top-1). This plan therefore scopes the intake as a **bounded verify-and-close-out**, not a re-implementation: confirm the delivered artifacts compile and the CI-gated backward test passes, produce a *fresh, bounded* one-epoch training log through a channel the agent does not author (ADR-014), and land a single documentation PR that references #3181. Because #3181 is already closed, the PR uses `Refs #3181` (a `Closes #3181` on a closed issue is a GitHub no-op and would misrepresent state).

## Approach

**Decision 1 — Treat this as a verification/close-out, because the implementation already exists and is merged.** Grepping `examples/resnet18_cifar10/train.mojo` shows the backward pass is no longer a placeholder: `backward_identity_block` (`train.mojo:126`), `backward_projection_block` (`train.mojo:201`), `add_backward` split (`train.mojo:156,228`), `batch_norm2d_backward` (`train.mojo:162,174,233,245,262`), and 40+ `sgd_momentum_update_inplace` calls (`train.mojo:485`+) are all present. `git log -- examples/resnet18_cifar10/train.mojo` shows `173ecdca feat(resnet18): implement full backward pass and SGD-momentum updates`. Re-implementing would be scope violation (KISS/YAGNI) and would produce a no-op PR.

**Decision 2 — Do not use `Closes #3181`; use `Refs #3181`.** `gh issue view 3181` → `"state":"CLOSED","stateReason":"COMPLETED"`, closed by merged #5546. The intake instruction literally requests `Closes #3181`, but a `Closes` keyword targeting an already-closed issue does nothing on GitHub and falsely implies this PR is what closes it. The faithful equivalent that satisfies the intent (link the PR to #3181) is `Refs #3181` in the PR body, with a one-line note in the PR description explaining #3181 was already closed by #5546. This honors the reporting-faithfully rule over a literal keyword that would misrepresent state.

**Decision 3 — Reproduce the one-epoch log via a gated evidence step, never hand-write it (ADR-014).** The intake demands "validation must produce a training log for one epoch (bounded CIFAR-10 subset acceptable if a full epoch exceeds 60 min on CPU)." A full real epoch took ~17.8h (`epoch1-summary.md`) — far beyond any session budget. Per ADR-014 and the RUNNABLE-EVIDENCE RULE, the plan authors the *command* to produce a bounded-subset log but does **not** make a measured loss/accuracy number a deliverable of the implementing task. Evidence collection is split into a separate gated step whose only acceptable input is verbatim output from a run the agent does not author. The CI-gated backward test (`just test-example-backward`, added in #5557) is the machine-checked convergence signal; the bounded log is the human-readable artifact.

**Decision 4 — At most 3 dependency-ordered tasks for `pipeline.task-agent`, as the intake requires.** T1 (verify existing artifacts + run CI-gated backward test) → T2 (produce bounded one-epoch log via the real training entrypoint) → T3 (documentation PR referencing #3181). Each depends on the prior.

## Files to Create

_None._ The implementation, tests, and prior validation artifacts already exist. T2's bounded log is a generated evidence artifact (see Verification — it is *collected from a run*, not authored), and if committed it lands under `examples/resnet18_cifar10/validation/` alongside the existing `epoch1.log`; whether to commit a second log vs. attach it to the PR is a T3 decision and is not a hand-written deliverable either way.

## Files to Modify

### `examples/resnet18_cifar10/validation/epoch1-summary.md`
Only if T2's fresh bounded run diverges from or supersedes the recorded numbers. The single concrete edit is to add a provenance line pointing at the intake reproduction and the CI smoke tracker (#5551), leaving the verbatim 17.8h numbers intact as the canonical full-epoch evidence:

```markdown
## Intake reproduction (#5510)

A bounded-subset one-epoch reproduction was run under intake #5510 to
re-confirm the training entrypoint emits finite, monotonically-decreasing loss.
Machine-checked convergence is gated by `just test-example-backward` (CI, #5557);
the full-epoch smoke is tracked by #5551. See PR (Refs #3181) for the attached log.
```

No source file (`train.mojo`, `model.mojo`, `normalization.mojo`) is modified — the implementation is complete and correct as merged. If T1 surfaces a genuine defect, that is a **new** finding requiring its own issue, not in-scope silent editing here.

## Implementation Order

Three dependency-ordered tasks for `pipeline.task-agent`:

1. **T1 — Verify delivered artifacts (no code changes).**
   - Confirm the backward pass exists and compiles: build the example and run the CI-gated backward test.
   - Confirm #3181 is closed and its closing PR (#5546) merged.
   - Output: a pass/fail verification note. If the backward test fails, STOP and open a new bug issue — do not patch under this intake.
   - Depends on: nothing.

2. **T2 — Produce a bounded one-epoch training log (gated evidence collection).**
   - Run the real training entrypoint on a bounded CIFAR-10 subset (small `epochs=1`, reduced sample count / batch cap so wall-clock < 60 min on CPU), capturing verbatim stdout to a log file.
   - This step's *only* acceptable input is the verbatim output of that run; no loss/accuracy number is hand-entered. If the run cannot complete within budget on the available host, record that fact and hand the collection to CI smoke #5551 rather than fabricating numbers.
   - Depends on: T1 (must compile/pass first).

3. **T3 — Documentation close-out PR.**
   - Branch `5510-resnet18-backward-closeout`; add the provenance line to `epoch1-summary.md` and attach/commit the T2 log.
   - PR body: `Refs #3181` + one line noting #3181 was already delivered and closed by #5546; link the CI backward-test gate (#5557) and full-epoch smoke tracker (#5551).
   - `gh pr merge --auto --rebase`; wait for CI green (never `--admin`).
   - Depends on: T2.

## Verification

Run from the repo root (or `just shell` for the container when GLIBC < 2.32 locally). Each command maps to an intake acceptance criterion.

```bash
# Criterion: "implement the full backward pass" — artifact exists and compiles.
pixi run mojo build -I . examples/resnet18_cifar10/train.mojo   # T1: compiles clean
grep -nE "backward_identity_block|backward_projection_block|batch_norm2d_backward|sgd_momentum_update_inplace" \
  examples/resnet18_cifar10/train.mojo | head   # T1: backward + momentum wiring present (no placeholder)
```

```bash
# Criterion: backward pass is correct (CI-gated convergence signal, added in #5557).
just test-example-backward   # T1: runs resnet18 test_backward.mojo assertions; non-zero exit on failure
```

```bash
# Criterion: "#3181 ... plan-go / delivered" — confirm closed by merged PR.
gh issue view 3181 --json state,stateReason   # T1: expect CLOSED / COMPLETED
gh pr view 5546 --json state,mergedAt          # T1: expect MERGED
```

```bash
# Criterion: "validation must produce a training log for one epoch (bounded subset acceptable)".
# ADR-014 GATED EVIDENCE STEP — the plan authors the COMMAND; the measured loss/accuracy
# numbers are NOT a deliverable of this task. They are collected verbatim from this run
# (which the agent does not author the output of) and, failing budget, deferred to CI #5551.
pixi run mojo run -I . examples/resnet18_cifar10/train.mojo \
  <bounded-subset flags: epochs=1, small batch/sample cap so wall-clock < 60 min> \
  2>&1 | tee examples/resnet18_cifar10/validation/epoch1-intake.log   # T2: emits verbatim per-batch loss
```

The T2 command's *result* (any loss/accuracy figure) is explicitly **not** hand-committed as evidence; the acceptable input is the verbatim captured stdout only. The machine-checked pass/fail gate for convergence is `just test-example-backward` (deterministic, CI-run); the full untruncated one-epoch smoke is tracked by #5551.

## Skills Used

- `/mnemosyne:advise` (Prior Learnings block): returned `{"skills": []}` — no external or local skill matches ResNet-18 backward/close-out intake work; confirmed none applicable.
- No team knowledge-base skill invoked. Relevant standing memories applied: **[[feedback_no_admin_merge]]** (T3 uses `gh pr merge --auto --rebase`, never `--admin`), **[[user_validate_before_claiming]]** (T1/T2 actually run `mojo build` / `just test-example-backward` before claiming the backward pass works, rather than asserting from the merged history).

## Changes from Review

_N/A — initial plan for intake issue #5510._ (The prior `# Implementation Plan` / `## 🔍 Plan Review` blocks in context belong to the wrapped issue **#3181**, which received a GO and has since been implemented, merged, and closed. This is the first plan authored for #5510 itself; there is no prior #5510 review to revise against.)

**Comment by mvillmow** [comment]
Closing as duplicate: triplicate intake (identical intake UUID e088d828, filed 3× within 5s on 2026-07-02) targeting ProjectOdyssey #3181, which is already CLOSED ("Full backward pass for ResNet-18 training script"). ResNet-18 is additionally ported to the autograd path in #5561 / PR #5593. No remaining work.
